src/mavsdk/plugins/component_metadata/file_cache.cpp

```diff
@@ -0,0 +1,231 @@
+
+#include "file_cache.h"
+#include "log.h"
+
+#include <cstring>
+#include <fstream>
+
+// For directory locking
+#ifdef _WIN32
+#include <Windows.h>
+#else
+#include <sys/file.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#endif
+
+namespace mavsdk {
+
+FileCache::DirectoryLock::DirectoryLock(const std::filesystem::path& path) : _path(path)
+{
+#ifdef _WIN32
+    _lock = CreateFileA(_path.c_str(), GENERIC_READ, FILE_SHARE_READ, 0, OPEN_EXISTING, 0, 0);
+    if (!LockFileEx(_lock, LOCKFILE_EXCLUSIVE_LOCK, 0, _lock_size, 0, &_overlapped)) {
+        LogErr() << "Cannot take lock file " << GetLastError();
+    }
+#else
+    mode_t m = umask(0);
+    _fd = open(_path.c_str(), O_RDWR | O_CREAT, 0666);
+    umask(m);
+    if (_fd < 0) {
+        LogErr() << "Cannot open file " << _path;
+    } else {
+        if (flock(_fd, LOCK_EX) == -1) {
+            LogErr() << "Cannot lock file " << strerror(errno);
```

Comment:
What happens in this case? It doesn't abort the program, right?

Reply:
No, it just continues. I thought that's ok given the chance for a race condition is small, and there is an error printed.

Reply:
Fair enough.